File: voice-agents/ios_swift_voice_agent/Sources/Clients/AtomsRest.swift
import Foundation

/// Thin wrapper around the Atoms REST surface the app needs to update a live
/// agent's voice / speed / language. Full dance (v2 branches flow):
///   1. GET  /agent/{id}                              read current
///   2. GET  /agent/{id}/branches                     find the live branch
///   3. PUT  /agent/{id}/branches/{b}/draft           write new values into the open draft
///   4. POST /agent/{id}/branches/{b}/draft/publish   publish; the revision goes live
enum AtomsRest {
    private static let base = "https://api.smallest.ai/atoms/v1"

    struct AgentSnapshot {
        var name: String
        var voiceId: String
        var voiceModel: String
        var speed: Double
        var language: String
        var supportedLanguages: [String]
    }

    struct UpdateInput {
        var voiceId: String?
        var voiceModel: String?
        var speed: Double?
        var language: String?
    }

    enum RestError: Error {
        case http(status: Int, body: String)
        case parse
        case noBranch
        case noRevisionId
    }

    static func fetchAgent(apiKey: String, agentId: String) async throws -> AgentSnapshot {
        let json = try await request(method: "GET", path: "/agent/\(agentId)", apiKey: apiKey)
        let data = (json["data"] as? [String: Any]) ?? json
        let synth = (data["synthesizer"] as? [String: Any]) ?? [:]
        let voiceConfig = (synth["voiceConfig"] as? [String: Any]) ?? [:]
        let language = (data["language"] as? [String: Any]) ?? [:]
        return AgentSnapshot(
            name: (data["name"] as? String) ?? "",
            voiceId: (voiceConfig["voiceId"] as? String) ?? "",
            voiceModel: (voiceConfig["model"] as? String) ?? "waves_lightning_v3_1",
            speed: (synth["speed"] as? Double) ?? 1.0,
            language: (language["default"] as? String) ?? "en",
            supportedLanguages: (language["supported"] as? [String]) ?? ["en"]
        )
    }

    @discardableResult
    static func updateAgentConfig(apiKey: String, agentId: String,
                                  current: AgentSnapshot, patch: UpdateInput) async throws -> String {
        let branchesJson = try await request(method: "GET",
                                             path: "/agent/\(agentId)/branches",
                                             apiKey: apiKey)
        let branchesData = (branchesJson["data"] as? [String: Any]) ?? branchesJson
        let branches = (branchesData["branches"] as? [[String: Any]]) ?? []
        // Pick the live branch; fall back to the default branch, then the first.
        let entry = branches.first { (($0["isLive"] ?? $0["is_live"]) as? Bool) == true }
            ?? branches.first { ((($0["branch"] as? [String: Any])?["isDefault"] ?? ($0["branch"] as? [String: Any])?["is_default"]) as? Bool) == true }
            ?? branches.first
        let branchObj = entry?["branch"] as? [String: Any]
        guard let branchId = (branchObj?["_id"] ?? branchObj?["id"]) as? String else {
            throw RestError.noBranch
        }

        let nextVoiceId     = patch.voiceId    ?? current.voiceId
        let nextVoiceModel  = patch.voiceModel ?? current.voiceModel
        let nextSpeed       = patch.speed      ?? current.speed
        let nextLanguage    = patch.language   ?? current.language

        var supported = current.supportedLanguages
        if !supported.contains(nextLanguage) { supported.append(nextLanguage) }

        let configBody: [String: Any] = [
            "language": [
                "default":   nextLanguage,
                "supported": supported,
                "switching": ["isEnabled": false],
            ],
            "synthesizer": [
                "voiceConfig": ["model": nextVoiceModel, "voiceId": nextVoiceId],
                "speed":       nextSpeed,
            ],
        ]
        // PUT creates the branch's open draft when there is none, otherwise
        // updates it in place.
        _ = try await request(method: "PUT",
                              path: "/agent/\(agentId)/branches/\(branchId)/draft",
                              apiKey: apiKey,
                              body: configBody)

        // Publishing the draft makes the new revision live; no activate step in v2.
        let publishJson = try await request(
            method: "POST",
            path: "/agent/\(agentId)/branches/\(branchId)/draft/publish",
            apiKey: apiKey,
            body: ["label": "ios-\(Int(Date().timeIntervalSince1970))"]
        )
        let publishData = (publishJson["data"] as? [String: Any]) ?? publishJson
        if let newRevision = (publishData["_id"] as? String) ?? (publishData["id"] as? String) {
            return newRevision
        }
        // Publishing runs an async security scan; poll until the draft closes.
        if publishData["state"] != nil {
            for _ in 0..<60 {
                try await Task.sleep(nanoseconds: 2_000_000_000)
                let branchJson = try await request(method: "GET",
                                                   path: "/agent/\(agentId)/branches/\(branchId)",
                                                   apiKey: apiKey)
                let data = (branchJson["data"] as? [String: Any]) ?? branchJson
                let branch = (data["branch"] as? [String: Any]) ?? data
                if branch["openDraftId"] == nil || branch["openDraftId"] is NSNull {
                    return (branch["headRevisionId"] as? String) ?? "published"
                }
            }
        }
        throw RestError.noRevisionId
    }

    // MARK: - Private

    private static func request(method: String, path: String, apiKey: String,
                                body: [String: Any]? = nil) async throws -> [String: Any] {
        guard let url = URL(string: base + path) else { throw RestError.parse }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.addValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        req.addValue("application/json", forHTTPHeaderField: "Content-Type")
        req.addValue("application/json", forHTTPHeaderField: "Accept")
        if let body {
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let (data, response) = try await URLSession.shared.data(for: req)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else {
            throw RestError.http(status: status, body: String(data: data, encoding: .utf8) ?? "")
        }
        if data.isEmpty { return [:] }
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw RestError.parse
        }
        return json
    }
}

File: voice-agents/react_native_voice_agent/src/agent/atomsRest.ts
// Thin wrapper around the Atoms REST surface the app needs for updating a
// live agent's voice/speed/language. Full dance (v2 branches flow):
//   1. GET /agent/{id}                              (read current config)
//   2. GET /agent/{id}/branches                     (find the live branch)
//   3. PUT /agent/{id}/branches/{b}/draft           (write new values into the open draft)
//   4. POST /agent/{id}/branches/{b}/draft/publish  (publish; the revision goes live)
// Anything that doesn't change is carried forward from the current config.

const API_BASE = 'https://api.smallest.ai/atoms/v1';

export interface AgentSnapshot {
  name: string;
  voiceId: string;
  voiceModel: string;
  speed: number;
  language: string;
  supportedLanguages: string[];
}

function headers(apiKey: string) {
  return {
    Authorization: `Bearer ${apiKey}`,
    'Content-Type': 'application/json',
    Accept: 'application/json',
  };
}

async function call<T>(
  apiKey: string,
  method: string,
  path: string,
  body?: unknown,
): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    method,
    headers: headers(apiKey),
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(`${method} ${path} → ${res.status}: ${text.slice(0, 240)}`);
  }
  if (res.status === 204) return {} as T;
  const text = await res.text();
  return text ? JSON.parse(text) : ({} as T);
}

function unwrap<T>(resp: any): T {
  return (resp && typeof resp === 'object' && 'data' in resp ? resp.data : resp) as T;
}

export async function fetchAgent(apiKey: string, agentId: string): Promise<AgentSnapshot> {
  const d = unwrap<any>(await call(apiKey, 'GET', `/agent/${agentId}`));
  return {
    name: d.name ?? '',
    voiceId: d?.synthesizer?.voiceConfig?.voiceId ?? '',
    voiceModel: d?.synthesizer?.voiceConfig?.model ?? 'waves_lightning_v3_1',
    speed: d?.synthesizer?.speed ?? 1.0,
    language: d?.language?.default ?? 'en',
    supportedLanguages: d?.language?.supported ?? ['en'],
  };
}

export interface UpdateInput {
  voiceId?: string;
  voiceModel?: string;
  speed?: number;
  language?: string;
}

// Runs the branch edit-and-publish flow. Returns the published revision id.
export async function updateAgentConfig(
  apiKey: string,
  agentId: string,
  current: AgentSnapshot,
  patch: UpdateInput,
): Promise<string> {
  const branchesResp = unwrap<any>(
    await call(apiKey, 'GET', `/agent/${agentId}/branches`),
  );
  const branches: any[] = branchesResp?.branches ?? [];
  const entry =
    branches.find((b) => b?.isLive ?? b?.is_live) ??
    branches.find((b) => b?.branch?.isDefault ?? b?.branch?.is_default) ??
    branches[0];
  const branchId: string = entry?.branch?._id ?? entry?.branch?.id;
  if (!branchId) throw new Error('No branch found on agent');

  const nextVoiceId = patch.voiceId ?? current.voiceId;
  const nextVoiceModel = patch.voiceModel ?? current.voiceModel;
  const nextSpeed = patch.speed ?? current.speed;
  const nextLanguage = patch.language ?? current.language;

  const configBody: Record<string, unknown> = {
    language: {
      default: nextLanguage,
      supported: current.supportedLanguages.includes(nextLanguage)
        ? current.supportedLanguages
        : Array.from(new Set([...current.supportedLanguages, nextLanguage])),
      switching: { isEnabled: false },
    },
    synthesizer: {
      voiceConfig: { model: nextVoiceModel, voiceId: nextVoiceId },
      speed: nextSpeed,
    },
  };

  // PUT creates the branch's open draft when there is none, otherwise
  // updates it in place.
  await call(apiKey, 'PUT', `/agent/${agentId}/branches/${branchId}/draft`, configBody);

  // Publishing the draft makes the new revision live; no activate step in v2.
  const publishResp = unwrap<any>(
    await call(apiKey, 'POST', `/agent/${agentId}/branches/${branchId}/draft/publish`, {
      label: `hearthside-${Date.now()}`,
    }),
  );
  let newRevision: string | undefined = publishResp?._id ?? publishResp?.id;
  if (!newRevision && publishResp?.state) {
    // Publishing runs an async security scan; poll until the draft closes.
    for (let i = 0; i < 60 && !newRevision; i++) {
      await new Promise((r) => setTimeout(r, 2000));
      const b = unwrap<any>(await call(apiKey, 'GET', `/agent/${agentId}/branches/${branchId}`));
      const branch = b?.branch ?? b;
      if (!branch?.openDraftId) newRevision = branch?.headRevisionId ?? 'published';
    }
  }
  if (!newRevision) throw new Error('Publish did not return revision id');

  return newRevision;
}
